### PR TITLE
Main helper

### PR DESCRIPTION
**Pull request by ash2k**

### CODE REVIEW

**Comment by wryun** [comment]
I'm dubious about whether this level of helping is worth it (I think ctrl is already over-reaching in terms of its integration point in some ways, and would be better if it acted as more of a libary), but I guess it's 'use it if you want to'.

**Comment by amckague** [comment]
example/main.go

**Comment by ash2k** [comment]
@wryun We need a pattern for controller startup and ideally as a reusable piece of code. We have this snippet replicated several times and I thought maybe it is worth extracting it. I'm ok if copy-pasting this bit is what we want to do.

My idea of what ctrl is: you should be able to use it as a library but also, if you want to, as a framework that takes over in some typical situations to reduce code-paste. That approach has been helpful because people were able to quickly start using it without copying and understanding the startup boilerplate. "Just copy this small snippet and plug your `Process()` function" (well, with an indirection for the controller constructor) is how I think about it. And then if/when this is not flexible enough, start using `ctrl` as a library and take the control back from it.

**Comment by ash2k** [comment]
An example of where the framework approach worked well is when we added some Prometheus metrics. Everyone got them for free without any work (the basic metrics). See https://github.com/atlassian/ctrl/pull/1/files#diff-5a6c724e55b16bbf8a5313965c60e181R81

**Comment by amckague** [comment]
What would it look like if this code as consumed in a project? (use an empty process function if you need to)

**Comment by ash2k** [comment]
```go
package main

imports (
    ctrlApp "ctrl/app"
)

func main() {
    ctrlApp.Main(innerMain)
}

func innerMain(ctx context.Context) error {
	// construct the controller, etc
	controllers := []ctrl.Constructor{
		&app.ControllerConstructor{},
	}
	a, err := ctrlApp.NewFromFlags(serviceName, controllers, flag.CommandLine, os.Args[1:])
	if err != nil {
		return err
	}
	return a.Run(ctx)
}
```

File: app/main.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+)
+
+func Main(main func(context.Context) error) {
+	if err := run(main); err != nil && errors.Cause(err) != context.Canceled {
```
Comment (scottgreenup) [inline app/main.go:12]: What about `context.DeadlineExceeded`? It's probably something we should echo out; just wondering if that was a choice or overlooked.

Comment (ash2k) [reply app/main.go:12]: It's on purpose. Program should not exit because of a timeout on a context. Not a controller for sure.

**Pull request closed without merging**